bridges/otellogr/logsink.go

```diff
@@ -0,0 +1,415 @@
+		return c
+	})
+}
+
+// NewLogSink returns a new [LogSink] to be used as a [logr.LogSink].
+//
+// If [WithLoggerProvider] is not provided, the returned LogSink will use the
+// global LoggerProvider.
+func NewLogSink(options ...Option) *LogSink {
+	c := newConfig(options)
+	return &LogSink{
+		config:        c,
+		logger:        c.logger(),
+		levelSeverity: c.levelSeverity,
+	}
+}
+
+// LogSink is a [logr.LogSink] that sends all logging records it receives to
+// OpenTelemetry. See package documentation for how conversions are made.
+type LogSink struct {
+	// Ensure forward compatibility by explicitly making this not comparable.
+	noCmp [0]func() //nolint: unused  // This is indeed used.
+
+	config        config
+	logger        log.Logger
+	levelSeverity func(int) log.Severity
+	values        []log.KeyValue
+}
+
+// Compile-time check *Handler implements logr.LogSink.
+var _ logr.LogSink = (*LogSink)(nil)
+
+// log sends a log record to the OpenTelemetry logger.
+func (l *LogSink) log(err error, msg string, serverity log.Severity, kvList ...any) {
+	var record log.Record
+	record.SetTimestamp(time.Now())
+	record.SetBody(log.StringValue(msg))
+	record.SetSeverity(serverity)
+
+	if err != nil {
+		record.AddAttributes(log.KeyValue{
+			Key:   errorKey,
+			Value: convertValue(err),
+		})
+	}
+
+	if len(l.values) > 0 {
+		record.AddAttributes(l.values...)
+	}
+
+	ctx, kv := convertKVs(kvList)
+	if len(kv) > 0 {
+		record.AddAttributes(kv...)
+	}
+
+	l.logger.Emit(ctx, record)
+}
+
+// Enabled tests whether this LogSink is enabled at the specified V-level.
+// For example, commandline flags might be used to set the logging
+// verbosity and disable some info logs.
+func (l *LogSink) Enabled(level int) bool {
+	var record log.Record
+	record.SetSeverity(l.levelSeverity(level))
+	ctx := context.Background()
+	return l.logger.Enabled(ctx, record)
+}
+
+// Error logs an error, with the given message and key/value pairs.
+func (l *LogSink) Error(err error, msg string, keysAndValues ...any) {
+	l.log(err, msg, log.SeverityError, keysAndValues...)
+}
+
+// Info logs a non-error message with the given key/value pairs.
+func (l *LogSink) Info(level int, msg string, keysAndValues ...any) {
+	l.log(nil, msg, l.levelSeverity(level), keysAndValues...)
+}
+
+// Init receives optional information about the logr library this
+// implementation does not use it.
+func (l *LogSink) Init(info logr.RuntimeInfo) {
+	// We don't need to do anything here.
+	// CallDepth is used to calculate the caller's PC.
+	// PC is dropped as part of the conversion to the OpenTelemetry log.Record.
+}
+
+// WithName returns a new LogSink with the specified name appended.
+func (l LogSink) WithName(name string) logr.LogSink {
+	newConfig := l.config
+	newConfig.scope.Name = fmt.Sprintf("%s/%s", l.config.scope.Name, name)
+
+	return &LogSink{
+		config:        newConfig,
+		logger:        newConfig.logger(),
+		levelSeverity: newConfig.levelSeverity,
+		values:        l.values,
+	}
```

Comment:
Couldn't we do the same thins as `WithValues` here, reusing the `LogSink` struct?